text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,284 @@
+# Resource and Entities - Data Model Part 2
+
+This is a proposal to address Resource and Entity data model interactions,
+including a path forward to address immediate friction and issues in the
+current resource specification.
+
+<!-- toc -->
+
+<!-- tocstop -->
+
+## Motivation
+
+This proposal attempts to focus on the following problems within OpenTelemetry to unblock multiple working groups:
+
+- Allowing mutating attributes to participate in Resource ([OTEP 208](https://github.com/open-telemetry/oteps/pull/208)).
+- Allow Resource to handle entities whose lifetimes don't match the SDK's lifetime ([OTEP 208](https://github.com/open-telemetry/oteps/pull/208)).
+- Provide support for async resource lookup ([spec#952](https://github.com/open-telemetry/opentelemetry-specification/issues/952)).
+- Fix current Resource merge rules in the specification, which most implementations violate ([oteps#208](https://github.com/open-telemetry/oteps/pull/208), [spec#3382](https://github.com/open-telemetry/opentelemetry-specification/issues/3382), [spec#3710](https://github.com/open-telemetry/opentelemetry-specification/issues/3710)).
+- Allow semantic convention resource modeling to progress ([spec#605](https://github.com/open-telemetry/opentelemetry-specification/issues/605), [spec#559](https://github.com/open-telemetry/opentelemetry-specification/issues/559), etc).
+
+## Design
+
+### Approach - Resource Improvements
+
+Let's focus on outlining Entity detectors and Resource composition.  This has a higher priority for fixing within OpenTelemetry, and needs to be unblocked sooner. Then infer our way back to data model and Collector use cases.
+
+We define the following SDK components:
+
+- **Resource Detectors (legacy)**:  We preserve existing resource detectors.  They have the same behavior and interfaces as today.
+- **Entity Detectors (new)**: Detecting an entity that is relevant to the current instance of the SDK.  For example, this would detect a service entity for the current SDK, or its process. Every entity must have some relation to the current SDK.
+- **Resource Coordinator (new)**: A component responsible for taking Resource and Entity detectors and doing the following:
+  - Constructing a Resource for the SDK from detectors.
+  - Dealing with conflicts between detectors.
+  - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.
```

Comment:
The TL;DR; here is that we agreed "Entity" signal would be an Event/Log.  The idea here is that you could export discovered Entities or entity change events via logs in the SDK here.

I can remove this as it's not that important, but something we can add later if we decide entities should be exported from SDKs + their detection.

Reply:
I assume using logs [like this](https://docs.google.com/document/d/1Tg18sIck3Nakxtd3TFFcIjrmRO_0GLMdHXylVqBQmJA/edit#heading=h.v4ilwdkncxe)?

Reply:
Yes.  I suspect we'll need to look at what cloud events is doing here as well, but I don't think we need this functionality in an initial implementation.  This sets up our architecture to support that feature (which I suspect we will need, but can defer).

Reply:
Second that this will be needed, as appending the full state of entities to all outgoing telemetry can be quite expensive, as can managing partial updates on the back-end. Creating a separate events stream with full entity state can relieve these pressures.

Reply:
@HunterSherms Absolutely agree.  The important point here is that this MAY happen directly from SDKs and not just within an OTEL collector, e.g.